### PR TITLE
Added in mmc5603 code

### PR DESCRIPTION
**Pull request by benhoff**
# What does this implement/fix?

Implements code for a MMC5603 magnet sensor

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2524

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
esphome:
  name: water

esp8266:
  board: d1_mini

# Enable logging
logger:

i2c:
  sda: D2
  scl: D1
  scan: True

wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

sensor:
  - platform: mmc5603
    address: 0x30
    field_strength_x:
      name: "HMC5883L Field Strength X"
    field_strength_y:
      name: "HMC5883L Field Strength Y"
    field_strength_z:
      name: "HMC5883L Field Strength Z"
    update_interval: 1s

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/mmc5603/sensor.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    def validate_enum_bound(value):
+        value = cv.string(value)
+        for unit in _units:
+            if value.endswith(unit):
+                value = value[: -len(unit)]
+                break
+        return enum_bound(value)
+
+    return validate_enum_bound
```
Comment (jesserockz) [inline esphome/components/mmc5603/sensor.py]: Not used anywhere?

File: esphome/components/mmc5603/mmc5603.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  void update() override;
+
+  void set_datarate(MMC5603Datarate datarate) { datarate_ = datarate; }
+  void set_x_sensor(sensor::Sensor *x_sensor) { x_sensor_ = x_sensor; }
+  void set_y_sensor(sensor::Sensor *y_sensor) { y_sensor_ = y_sensor; }
+  void set_z_sensor(sensor::Sensor *z_sensor) { z_sensor_ = z_sensor; }
+  void set_heading_sensor(sensor::Sensor *heading_sensor) { heading_sensor_ = heading_sensor; }
+
+ protected:
+  MMC5603Datarate datarate_{MMC5603_DATARATE_150_0_HZ};
```
Comment (jesserockz) [inline esphome/components/mmc5603/mmc5603.h]: Value is always set from config/codegen
```suggestion
  MMC5603Datarate datarate_;
```

File: esphome/components/mmc5603/mmc5603.cpp [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  LOG_SENSOR("  ", "Y Axis", this->y_sensor_);
+  LOG_SENSOR("  ", "Z Axis", this->z_sensor_);
+  LOG_SENSOR("  ", "Heading", this->heading_sensor_);
+}
+
+float MMC5603Component::get_setup_priority() const { return setup_priority::DATA; }
+
+void MMC5603Component::update() {
+  // check if continuous mode first
+  // if (!isContinuousMode())
```
Comment (jesserockz) [inline esphome/components/mmc5603/mmc5603.cpp]: No commented code please
```suggestion
```

File: esphome/components/mmc5603/mmc5603.cpp [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    this->status_set_warning();
+    return;
+  }
+  uint8_t status = 0;
+  if (!this->read_byte(MMC56X3_STATUS_REG, &status)) {
+    this->status_set_warning();
+    return;
+  }
+
+  status = status << 6;
```
Comment (jesserockz) [inline esphome/components/mmc5603/mmc5603.cpp]: Never used again?
```suggestion
```

**Review by jesserockz — CHANGES_REQUESTED** [review]
Minor changes, otherwise fine

**Comment by benhoff** [comment]
@jesserockz , do I need to do anything like bump this? Should I let the bot close it as stale? I am using this PR for one of my devices.

**Review by jesserockz — APPROVED** [review]

**Comment by jesserockz** [comment]
Sorry for not getting to this earlier.

**Pull request merged**